### PR TITLE
Feature/height chart

### PR DESCRIPTION
**Pull request by LB31**
- Chart needs routes to be tested
- ElevationGetter needs to be awaited when used
- ElevationChart currently shows the amount of nodes instead of the route distance on the x-axis

### CODE REVIEW

**Comment by LB31** [comment]
- added descriptions for axes
- made descriptions and chart interaction optional
- the x-axis shows know the actual distance

File: .flutter-plugins-dependencies [hunk 1]
```diff
@@ -0,0 +1 @@
+{"_info":"// This is a generated file; do not edit or check into version control.","dependencyGraph":[{"name":"esys_flutter_share","dependencies":["path_provider"]},{"name":"geocoder","dependencies":[]},{"name":"location","dependencies":[]},{"name":"location_permissions","dependencies":[]},{"name":"mapbox_gl","dependencies":[]},{"name":"path_provider","dependencies":[]},{"name":"share","dependencies":[]},{"name":"shared_preferences","dependencies":[]},{"name":"sqflite","dependencies":[]}]}
```
Comment (dephiloper) [inline .flutter-plugins-dependencies:1]: .flutter-plugins-dependencies should be ignored see line 47. https://github.com/flutter/flutter/blob/master/.gitignore

please put this into the gitignore

File: android/gradle.properties [hunk 2]
```diff
@@ -1,3 +1,4 @@
 org.gradle.jvmargs=-Xmx1536M
 android.enableJetifier=true
 android.useAndroidX=true
+android.enableR8=true
```
Comment (dephiloper) [inline android/gradle.properties:4]: what does this option do? Is this really necessary?

File: lib/pages/home.dart [hunk 3]
```diff
@@ -2,18 +2,33 @@ import 'package:flutter/material.dart';
...
 import 'package:hiking4nerds/components/navbar.dart';
 import 'package:hiking4nerds/components/shareroute.dart';
+import 'package:hiking4nerds/services/ElevationChart.dart';
+import 'package:hiking4nerds/services/elevationQuery.dart';
 import 'package:hiking4nerds/services/osmdata.dart';
 import 'package:hiking4nerds/services/route.dart';
 import 'package:hiking4nerds/styles.dart';
 
 class Home extends StatefulWidget {
+  static bool chartIsHere = false;
```
Comment (dephiloper) [inline lib/pages/home.dart]: this is only used for an print so I guess that could be removed

File: lib/pages/home.dart [hunk 4]
```diff
@@ -2,18 +2,33 @@ import 'package:flutter/material.dart';
...
   @override
   _HomeState createState() => _HomeState();
 }
 
 class _HomeState extends State<Home> {
   @override
   Widget build(BuildContext context) {
+    bool done = false;
```
Comment (dephiloper) [inline lib/pages/home.dart]: done gets set but is never used, is there a reason to have this variable?

File: lib/pages/home.dart [hunk 5]
```diff
@@ -2,18 +2,33 @@ import 'package:flutter/material.dart';
...
+    HikingRoute route = new HikingRoute([
+      Node(0, 52.510318, 13.4085592),
+      Node(1, 52.5102903, 13.4084606),
+      Node(2, 52.5101514, 13.4081806),
+      Node(3, 52.507592, 13.409908),
+    ], 5);
+    ElevationGetter.queryElevations(route).then((value) {
+      done = true;
+      route.elevations = value;
```
Comment (dephiloper) [inline lib/pages/home.dart]: value is a weak name for this parameter here, elevations would be better

File: lib/pages/home.dart [hunk 6]
```diff
@@ -22,16 +37,20 @@ class _HomeState extends State<Home> {
...
           Align(
             alignment: Alignment.bottomRight,
             child: RawMaterialButton(
               onPressed: () {
-                HikingRoute mockRoute = HikingRoute([
-                  Node(0, 52.510318, 13.4085592),
-                  Node(1, 52.5102903, 13.4084606),
-                  Node(2, 52.5101514, 13.4081806)
-                ], 2);
+                HikingRoute mockRoute = route;
```
Comment (dephiloper) [inline lib/pages/home.dart]: declarated mockroute only once at the top of this file and use it everywhere, this copy isn't really needed

File: lib/pages/settings.dart [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+class Settings extends StatefulWidget {
+  @override
+  _SettingsState createState() => _SettingsState();
+}
+
+class _SettingsState extends State<Settings> {
+  @override
+  Widget build(BuildContext context) {
+    print("CHART IS HERE2 " + Home.chartIsHere.toString());
```
Comment (dephiloper) [inline lib/pages/settings.dart]: please remove the print
why is the settings page created in this branch when the branch is for the height chart only?

File: lib/services/ElevationChart.dart [hunk 8]
```diff
@@ -0,0 +1,130 @@
+import 'package:charts_flutter/flutter.dart' as charts;
+import 'package:flutter/material.dart';
+import 'package:hiking4nerds/services/route.dart';
+
+import 'osmdata.dart';
+
+class ElevationChart extends StatelessWidget {
```
Comment (dephiloper) [inline lib/services/ElevationChart.dart]: naming the class ElevationChart is fine but the file should be not be named in CamelCase
right naming would be snake_case 
see: https://dart.dev/guides/language/effective-dart/style

furthermore elevationchart is an widget and should be inside compontents not under service

File: lib/services/ElevationChart.dart [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+  final List<charts.Series> seriesList;
+  final bool interactive;
+  final bool withLabels;
+
+  int index;
```
Comment (dephiloper) [inline lib/services/ElevationChart.dart]: Flutter marks here that you have specified that this widget is stateless, but then on the other hand you have defined an 'index' here which gets modified and stored while using the chart. Therefore this widget should have an state (extending from StatefulWidget).

Probably the index variable isn't needed at all, if this is true delete it and let the widget stay stateless.

File: lib/services/ElevationChart.dart [hunk 10]
```diff
@@ -0,0 +1,130 @@
+import 'package:charts_flutter/flutter.dart' as charts;
+import 'package:flutter/material.dart';
+import 'package:hiking4nerds/services/route.dart';
+
+import 'osmdata.dart';
+
+class ElevationChart extends StatelessWidget {
+  final List<charts.Series> seriesList;
```
Comment (dephiloper) [inline lib/services/ElevationChart.dart]: why should the elevation chart get an list of series handed over? shouldn't this be an route object?

File: lib/services/ElevationChart.dart [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+          color: Colors.black,
+          width: 1,
+        ),
+        borderRadius: BorderRadius.circular(6),
+      ),
+    );
+  }
+
+  /// Create one series with sample hard coded data.
+  static List<charts.Series<RouteChart, double>> _createData(HikingRoute route,
```
Comment (dephiloper) [inline lib/services/ElevationChart.dart]: i guess this is only for testing? if not change naming of the variables: lastX, x, sales, 

I think to test the chart properly you should only have to pass the mockroute into it, no need for creating additional mock data in the widget.

File: lib/services/ElevationChart.dart [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+import 'osmdata.dart';
+
+class ElevationChart extends StatelessWidget {
+  final List<charts.Series> seriesList;
+  final bool interactive;
+  final bool withLabels;
+
+  int index;
+
+  ElevationChart(this.seriesList, {this.interactive, this.withLabels});
```
Comment (dephiloper) [inline lib/services/ElevationChart.dart]: the callback function is missing here for us to insert as parameter.

when creating an elevation chart there should be the option to set an callback as parameter which gets called when the user drags over the points of the elevation chart. You have created an function inside this widget but it does not call the callback function therefore there is no access from outside the widget.

parent_page ---- callback function ----> elevation_chart

elevation_chart->user_drag_event() { call callback function with index as parameter }

File: lib/services/osmdata.dart [hunk 13]
```diff
@@ -212,7 +212,7 @@ class OsmData{
     }
   }
 
-  static double getDistance(LatLng nodeA, LatLng nodeB){
+  static double getDistance(nodeA, nodeB){
```
Comment (dephiloper) [inline lib/services/osmdata.dart]: why is the type removed here. This should not be necessary anymore.

File: lib/services/route.dart [hunk 14]
```diff
@@ -1,13 +1,14 @@
...
-class HikingRoute {
 
+class HikingRoute {
   List<Node> path; //path of route
   double totalLength; // routeLength in km
   List<PointOfInterest> pointsOfInterest;
   //todo: add height information
+  List<double> elevations; // elevations of route points in m; route points and their elevations have the same index
+
+  HikingRoute(this.path, this.totalLength,[this.pointsOfInterest]);
```
Comment (dephiloper) [inline lib/services/route.dart]: why not adding the elevation to as optional parameter to the constructor

File: lib/services/ElevationChart.dart [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+  /// Create one series with sample hard coded data.
+  static List<charts.Series<RouteChart, double>> _createData(HikingRoute route,
+      [bool interactive, bool withLabels]) {
+    //route.elevations = [3.3, 2.1, 50.2, 20.8]; // TODO remove, just for testing
+
+    final List<RouteChart> chartData = new List();
+
+    double lastX = 0;
+
+    for (var i = 0; i < route.elevations.length; i++) {
```
Comment (dephiloper) [inline lib/services/ElevationChart.dart]: types are prefered, we suggested in one meeting to use them

File: lib/services/elevationQuery.dart [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+class ElevationGetter {
+
+  static Future<List<double>> queryElevations(HikingRoute route) async {
+    List<Node> path = route.path;
+    List<double> queriedElevations = new List();
+    String basisURL = "https://h4nsolo.f4.htw-berlin.de/elevation/api/v1/lookup?locations=";
+    String currentURL = basisURL;
+    int nodesPerQuery = 100; // they have to be devided because of the maximum URL character amount
+
+    for (var i = 0; i < path.length; i++) {
```
Comment (dephiloper) [inline lib/services/elevationQuery.dart]: types are prefered, we suggested in one meeting to use them

File: lib/services/elevationQuery.dart [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+      currentURL += path[i].latitude.toString() + "," + path[i].longitude.toString();
+
+      if (i % nodesPerQuery == 0 && i > 0 || i == path.length - 1) { // every nodesPerQuery steps or when all paths were visisted
+        var response = await http.get(currentURL);
```
Comment (dephiloper) [inline lib/services/elevationQuery.dart]: types are prefered, we suggested in one meeting to use them

File: pubspec.yaml [hunk 18]
```diff
@@ -29,6 +29,8 @@ dependencies:
   fluttertoast: ^3.1.3
   geocoder: ^0.2.1
   shared_preferences: ^0.5.4
+  charts_flutter: ^0.8.1
+  fl_chart: ^0.6.0
```
Comment (dephiloper) [inline pubspec.yaml]: fl_chart is not needed

File: lib/services/ElevationChart.dart [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+  final List<charts.Series> seriesList;
+  final bool interactive;
+  final bool withLabels;
+
+  int index;
+
+  ElevationChart(this.seriesList, {this.interactive, this.withLabels});
+
+  /// Creates a [LineChart] with sample data and no transition.
+  factory ElevationChart.withData(HikingRoute route,
```
Comment (dephiloper) [inline lib/services/ElevationChart.dart]: the factory is one way, but i think the more intuitve way would be to set the route via the constructor (@required) and use the route data inside build to create the chart correctly

**Review by dephiloper — CHANGES_REQUESTED** [review]
There are some minor and some major issues that you should consider to change. Especially the index should be removed imo and you should set the route and an callback function as constructor params.

File: android/gradle.properties (hunk 2 above)
Comment (LB31) [reply android/gradle.properties:4]: https://android-developers.googleblog.com/2018/11/r8-new-code-shrinker-from-google-is.html

seems to be a code shrinking feature for Android builds. Was inserted automatically, but sounds not wrong for me.

File: lib/pages/home.dart (hunk 4 above)
Comment (LB31) [reply lib/pages/home.dart]: This one was experimental. I will remove it.
In the future we need to see how code parts that are pulling elevations should handle the waiting without disturbing the rest of the flow.

File: lib/services/ElevationChart.dart (hunk 12 above)
Comment (LB31) [reply lib/services/ElevationChart.dart]: see onSelectionChanged parameter in the next push

**Comment by LB31** [comment]
@dephiloper thank you for your detailed review. The issues should be solved now.

File: android/gradle.properties (hunk 2 above)
Comment (dephiloper) [reply android/gradle.properties:4]: alright

**Review by dephiloper — APPROVED** [review]

**Pull request merged**